src/dupin.h

```diff
@@ -0,0 +1,144 @@
+#pragma once
+
+#include <algorithm>
+#include <iostream>
+#include <limits>
+#include <unordered_map>
+#include <vector>
+#include <Eigen/Dense>
+
+
+// DynamicProgramming class for dynamic programming based segmentation.
+class DynamicProgramming { 
+private:
+  class UpperTriangularMatrix {
+  private:
+    std::vector<double> matrix;
+    std::vector<int> row_indices;
+    int length;
+
+    int index(int row, int col) const {
+        return row_indices[row] + col - row;
+    }
+
+  public:
+    UpperTriangularMatrix() : length(0) {}
+
+    UpperTriangularMatrix(int n) : length(n), matrix(n * (n + 1) / 2, 0.0), 
+                                   row_indices(n) {
+        for (int row = 0; row < n; ++row) {
+            row_indices[row] = row * (2 * length - row + 1) / 2;
+        }
+    }
+
+    void initialize(int n) {
+        length = n;
+        matrix.resize(n * (n + 1) / 2, 0.0);
+        row_indices.resize(n);
+        for (int row = 0; row < n; ++row) {
+            row_indices[row] = row * (2 * length - row + 1) / 2;
+        }
+    }
+
+    double &operator()(int row, int col) {
+        return matrix[index(row, col)];
+    }
+    int getSize() const { return length; }
+};
+  UpperTriangularMatrix cost_matrix;
+
+  // Struct for memoization key, combining start, end, and number of
+  // breakpoints.
+  struct MemoKey {
+    int start;
+    int end;
+    int num_bkps;
+
+    // Comparison operator for MemoKey.
+    bool operator==(const MemoKey &other) const {
+      return start == other.start && end == other.end &&
+             num_bkps == other.num_bkps;
+    }
+  };
+
+  // Custom XOR-bit hash function for MemoKey, avoids clustering of data in
+  // unordered map to improve efficiency.
+  struct MemoKeyHash { 
+    std::size_t operator()(const MemoKey &key) const {
+      return ((std::hash<int>()(key.start) ^
+               (std::hash<int>()(key.end) << 1)) >>
+              1) ^
+             std::hash<int>()(key.num_bkps);
+    }
+  };
+
+  // Memoization map to store the cost and partition for given parameters.
+  std::unordered_map<MemoKey, std::pair<double, std::vector<int>>, MemoKeyHash>
+      memo;
+
+  int num_bkps;          // Number of breakpoints to detect.
+  int num_parameters;    // Number of features in the dataset.
+  int num_timesteps;     // Number of data points (time steps).

```

Comment:
same as above^ global variables are used in a lot of the functions rather than data.col or data.rows.. if it's a matter of abstraction I can change it but it does help the code look more understandable

Reply:
Let me know if this is still an issue.. it will just take a bit of restructuring but I used these parameters throughout my functions rather than using signal.size etc. If it's all private/protected i'd imagine it wouldn't be an issue?